gcp_variant_transforms/beam_io/vcf_parser.py

```diff
@@ -504,6 +508,162 @@ def _get_variant_calls(self, record, formats):
     return calls
 
 
+class PySamParser(VcfParser):
+  """An Iterator for processing a single VCF file using PySam.
+
+  PySam allows reading a file through either stdin stream, or through as actual
+  VCF files, for which it requires legitimate file descriptor. Since we want to
+  perform our own parallelization, we will fork our process, to use 2 pipelines
+  that will feed into PySam object - 1 will feed the data from main thread into
+  the child one, while second will get that data in child thread and feed it to
+  PySam library.
+
+  The requirement of using two pipelines comes from the design of VcfParser base
+  class - we could only use a single pipe, but it will divert the parsers.
+  """
+
+  def __init__(self,
+               file_name,  # type: str
+               range_tracker,  # type: range_trackers.OffsetRangeTracker
+               compression_type,  # type: str
+               allow_malformed_records,  # type: bool
+               file_pattern=None,  # type: str
+               representative_header_lines=None,  # type:  List[str]
+               splittable_bgzf=False,  # type: bool
+               **kwargs  # type: **str
+              ):
+    # type: (...) -> None
+    super(PySamParser, self).__init__(file_name,
+                                      range_tracker,
+                                      file_pattern,
+                                      compression_type,
+                                      allow_malformed_records,
+                                      representative_header_lines,
+                                      **kwargs)
+    self._header_lines = []
+    self._next_line_to_process = None
+    self._current_line = None
+    # This member will be properly initiated in _init_with_header().
+    self._vcf_reader = None
+    self._to_child = None
+
+  def _init_with_header(self, header_lines):
+    # The first header line must be similar to '##fileformat=VCFv.*'.
+    if header_lines and not header_lines[0].startswith(
+        FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='')):
+      header_lines.insert(0, FILE_FORMAT_HEADER_TEMPLATE.format(VERSION='4.0'))
+
+    p_read, c_write = os.pipe()
+    c_read, p_write = os.pipe()
+
+    processid = os.fork()
+    if processid:
+      os.close(c_read)
+      os.close(c_write)
+      into_pysam = os.fdopen(p_read)
+      self._to_child = os.fdopen(p_write, 'w')
+      self._vcf_reader = libcbcf.VariantFile(into_pysam, 'r')
```

Comment:
I don't know why we cannot use the file name directly to create the `VariantFile`.

Reply:
Which file name?

So as discussed offline, `VariantFile` only accepts actual file descriptor, and mocking it is not an option. Therefore, we create pipe to push data in. To use pipe, we need to create child subthread (in the else clause). Ideally, I would supply the entire record list through child thread and write it into the pipe, to get it parsed from VariantFile. However, due to the nature of how `vcf_parser` is structured, the `next()` reads the `text_lines` and parses them line by line, and unless I override it, I can only add logic in the `get_variant()`, where I parse the records on 1-by-1 basis. To achieve this, I create a second pipe:

Now, after supplying all of the header lines into the pipe, child thread waits for new records to come. When `get_variant()` is invoked with a record in the parent thread, it sends the record to child thread via second pipe. Once received, child pushes the record into the first pipe to the VariantFile class in parent thread, from where get_variant reads it back and returns it.

Reply:
Hmmm, doing the above refactoring is not difficult, but seems a bit pointless - those methods would have like 1, maybe 2 line each, and would not be repeated anywhere.

Instead, parsed out the child thread into its own method, and added lots of comments. WDYT?

Reply:
I generally agree with Allie here that it would be nicer if we had

if processid:
  writer()
else:
  reader()

sort of thing.